### PR TITLE
Interactive frontend

### PR DESCRIPTION
**Pull request by kdzwinel**
UI and functionality isn't 100% finished - this is a first PR. It will crate a separate page (for now) with two components: avatar and bubble.

pages-interactive mov

### CODE REVIEW

File: .babelrc [hunk 1]
```diff
@@ -1,4 +1,4 @@
 {
-  "presets": [["es2015"], "react"],
```
Comment (matikrk) [inline .babelrc]: Do we care about pages with don't support ES6? If we delete es2015 we will have in code arrow functions, const, let etc.
@aju @vergilius  WDYF?

File: src/components/avatar/pages/avatar-interactive.jsx [hunk 2]
```diff
@@ -0,0 +1,24 @@
+import React from 'react';
```
Comment (kdzwinel) [inline src/components/avatar/pages/avatar-interactive.jsx]: that's how doc section for Avatars will look when we switch to interactive components

File: src/components/action-list/pages/action-list.jsx [hunk 3]
```diff
@@ -5,8 +5,8 @@ import ButtonPrimary, {TYPE as BUTTON_PRIMARY_TYPE} from '../../buttons/ButtonPr
 import Label, {SIZE as LABEL_SIZE, ICON_COLOR, ICON_TYPE} from '../../labels/Label';
 import Icon from '../../icons/Icon';
 import Text, {TYPE as TEXT_TYPE, SIZE as TEXT_SIZE, COLOR as TEXT_COLOR, WEIGHT as TEXT_WEIGHT} from '../../text/Text';
-import ContrastBox from '../../../docs/ContrastBox';
-import DocsBlock from '../../../docs/DocsBlock';
+import ContrastBox from '../../../docs/components/ContrastBox';
+import DocsBlock from '../../../docs/components/DocsBlock';
```
Comment (kdzwinel) [inline src/components/action-list/pages/action-list.jsx]: just to clean things up I moved documentation specific components to one folder

File: src/docs/components/DocsActiveBlock.jsx [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+    const props = this.state.props;
+    props[key] = value;
+
+    this.setState({
+      props
+    })
+  }
+
+  render() {
+    const component = React.cloneElement(this.props.children, this.state.props);
```
Comment (kdzwinel) [inline src/docs/components/DocsActiveBlock.jsx]: creates a new version of the component with props from the Settings from

File: src/docs/components/ComponentSettings.jsx [hunk 5]
```diff
@@ -0,0 +1,88 @@
+import React from 'react';
```
Comment (kdzwinel) [inline src/docs/components/ComponentSettings.jsx]: Creates Settings form based on component props

File: src/docs/components/JSXGenerator.js [hunk 6]
```diff
@@ -0,0 +1,45 @@
+function isComponent(value) {
```
Comment (kdzwinel) [inline src/docs/components/JSXGenerator.js]: generates JSX from a component ✨

File: src/docs/components/CodeBlock.jsx [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+
+  let markup = beautify(children, {
+    indent_size: 2,
+    unformatted: [],
+    wrap_line_length: 0
+  });
+
+  if(type === 'jsx') {
+    //HACK <i> was added to force highlightJS to highlight first tag
+    markup = hljs.highlight('jsx', `<i>${markup}</i>`).value.slice(9, -78);
```
Comment (matikrk) [inline src/docs/components/CodeBlock.jsx]: magic numbers

File: src/docs/components/ComponentBlock.jsx [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+  constructor(props) {
+    super(props);
+    this.oldHTML = '';
+  }
+
+  componentDidMount() {
+    this.updateHTML();
+  }
+
+  componentDidUpdate() {
```
Comment (matikrk) [inline src/docs/components/ComponentBlock.jsx]: We can do it on `componentWillReceiveProps(nextProps)`

File: src/docs/components/ComponentBlock.jsx [hunk 9]
```diff
@@ -0,0 +1,44 @@
+import React, {Component} from 'react';
+import PropTypes from 'prop-types';
+
+class ComponentBlock extends Component {
+
+  constructor(props) {
+    super(props);
+    this.oldHTML = '';
```
Comment (matikrk) [inline src/docs/components/ComponentBlock.jsx]: put this on this.state.oldHTML

File: src/docs/components/ComponentBlock.jsx (hunk 8 above)
Comment (matikrk) [inline src/docs/components/ComponentBlock.jsx]: We can go even step father in this case

Do it on componentWillReceiveProps, delete oldHTML or even state.oldHTML
compare props if there are equal (equal reference) if not check if equal inner html by: ReactDOMServer.renderToStaticMarkup method. 

In that case we could store this.oldHTML for caching only, for faster calculating.

File: src/docs/components/ComponentSettings.jsx [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+    } else if (type === 'array') {
+      value = values[target.value];
+    } else if (type === 'string') {
+      value = target.value;
+    }
+
+    onChange(key, value);
+  }
+
+  const content = Object.keys(settings).map(key => {
```
Comment (matikrk) [inline src/docs/components/ComponentSettings.jsx]: Can we move it to separated component? And even 3 component, arrayComnponent, normalComponent

File: src/docs/components/DocsActiveBlock.jsx [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+      </DocsBlock>
+      <DocsBlock info="HTML">
+        <CodeBlock type="html">{html}</CodeBlock>
+      </DocsBlock>
+    </div>;
+  }
+}
+
+DocsActiveBlock.propTypes = {
+  children: PropTypes.node,
```
Comment (matikrk) [inline src/docs/components/DocsActiveBlock.jsx]: `children: PropTypes.element.isRequired` coz we have always 1 node and we take from it props

File: src/docs/components/JSXGenerator.js [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+  return '{' + JSON.stringify(prop) + '}';
+}
+
+function generateJSX(component) {
+  if (!isComponent(component)) {
+    return component;
+  }
+
+  const type = component.type.name || component.type;
+  let children = component.props.children || [];
```
Comment (matikrk) [inline src/docs/components/JSXGenerator.js]: I'm not sure if you doing it correctly
mapping to array should be done by React.Children.toArray(children)

more:
https://facebook.github.io/react/docs/react-api.html#react.children

**Comment by anna-wro** [comment]
Wow 😮 It's awesome!

File: .babelrc (hunk 1 above)
Comment (kdzwinel) [inline .babelrc]: tests do not run w/o "es2015" for some reason - I'm reverting that

File: src/docs/components/DocsActiveBlock.jsx [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+    this.setState({
+      props
+    });
+  }
+
+  render() {
+    const component = React.cloneElement(this.props.children, this.state.props);
+
+    const jsx = generateJSX(component);
+    const html = ReactDOMServer.renderToStaticMarkup(component);
```
Comment (matikrk) [inline src/docs/components/DocsActiveBlock.jsx]: 👍

File: src/components/layout/pages/layout.jsx [hunk 14]
```diff
@@ -1,5 +1,5 @@
 import React from 'react';
-import DocsBlock from '../../../docs/DocsBlock';
+import DocsBlock from '../../../docs/components/DocsBlock';
```
Comment (anna-wro) [inline src/components/layout/pages/layout.jsx]: `../../../`

File: src/components/avatar/pages/avatar.jsx [hunk 15]
```diff
@@ -1,7 +1,7 @@
 import React from 'react';
 import Avatar, {SIZE} from '../Avatar';
-import DocsBlock from 'DocsBlock';
-import ContrastBox from 'ContrastBox';
+import DocsBlock from '../../../docs/components/DocsBlock';
+import ContrastBox from '../../../docs/components/ContrastBox';
```
Comment (anna-wro) [inline src/components/avatar/pages/avatar.jsx]: `../../../` ;)

File: src/components/layout/pages/layout.jsx (hunk 14 above)
Comment (kdzwinel) [inline src/components/layout/pages/layout.jsx]: oh, thanks!

File: src/docs/components/ComponentSettings.jsx (hunk 10 above)
Comment (kdzwinel) [inline src/docs/components/ComponentSettings.jsx]: I'm leaving that for future PRs - Settings will be heavily reafactored anyway

File: scripts/tasks/build-react-interactive-pages.js [hunk 16]
```diff
@@ -0,0 +1,67 @@
...
+  module: {
+    loaders: [
+      {
+        test: /\.js|jsx?$/,
+        loader: 'babel-loader'
+      }
+    ]
+  },
+  externals: {
+    'html_beautify': 'html_beautify',
```
Comment (matikrk) [inline scripts/tasks/build-react-interactive-pages.js]: 👍

**Pull request merged**